File: iOS/AUiKit/AUiKit/Classes/Utils/RtmHelper/AUiRtmManager.swift
//
//  AUiRtmManager.swift
//  AUiKit
//
//  Created by wushengtao on 2023/3/1.
//

import Foundation
import AgoraRtcKit

let kChannelType = AgoraRtmChannelType.stream


/// 对RTM相关操作的封装类
open class AUiRtmManager: NSObject {
    private var streamChannel: AgoraRtmStreamChannel?
    private let proxy: AUiRtmMsgProxy = AUiRtmMsgProxy()
    
    private var rtmClient: AgoraRtmClientKit!
    private var rtmStreamChannelMap: [String: AgoraRtmStreamChannel] = [:]
    
    public private(set) var isLogin: Bool = false
    
    deinit {
        aui_info("deinit AUiRtmManager", tag: "AUiRtmManager")
    }
    
    init(rtmClient: AgoraRtmClientKit?) {
        self.rtmClient = rtmClient
        super.init()
        //TODO: 如果后续外部再修改rtmClient的delegate会有问题
        proxy.origRtmDelegate = self.rtmClient.agoraRtmDelegate
        self.rtmClient.agoraRtmDelegate = proxy
        aui_info("init AUiRtmManager", tag: "AUiRtmManager")
    }
    
    public func login(token: String, completion: @escaping (Error?)->()) {
        if isLogin {
            completion(nil)
            return
        }
        let ret =
        self.rtmClient.login(byToken: token) {[weak self] resp, error in
            aui_info("login: \(resp) \(error.errorCode.rawValue)", tag: "AUiRtmManager")
            self?.isLogin = error.errorCode == .ok ? true : false
            completion(error.toNSError())
        }
        
        //已经登陆过
        if abs(ret) == AgoraRtmErrorAlreadyLogin.rawValue {
            isLogin = true
            completion(nil)
        }
        
        aui_info("login ret: \(ret)", tag: "AUiRtmManager")
    }
    
    public func logout() {
        aui_info("logout", tag: "AUiRtmManager")
        rtmClient.logout()
        isLogin = false
    }
    
    public func renew(token: String) {
        aui_info("renew: \(token)", tag: "AUiRtmManager")
        rtmClient.renewToken(token)
    }
    
    public func renewChannel(channelName: String, token: String) {
        guard let streamChannel = rtmStreamChannelMap[channelName] else {
            return
        }
        
        aui_info("renewChannel: \(channelName) token: \(token)", tag: "AUiRtmManager")
        streamChannel.renewToken(token)
    }
}

//MARK: user
extension AUiRtmManager {
    public func getUserCount(channelName: String, completion:@escaping (Error?, Int)->()) {
        guard let presence = rtmClient.getPresence() else {
            completion(AUiCommonError.rtmError(-1).toNSError(), 0)
            return
        }
        
        let options = AgoraRtmPresenceOptions()
        options.includeUserId = false
        options.includeState = false
        let ret =
        presence.whoNow(channelName, channelType: kChannelType, options: options, completion: { resp, error in
//            aui_info("presence whoNow '\(channelName)' finished: \(error.errorCode.rawValue) list count: \(resp.userStateList.count) userId: \(AUiRoomContext.shared.commonConfig?.userId ?? "")", tag: "AUiRtmManager")
            aui_info("getUserCount: \(resp.totalOccupancy)", tag: "AUiRtmManager")
            let userList = resp.userList()
            completion(error.toNSError(), userList.count)
        })
        aui_info("presence whoNow '\(channelName)' ret: \(ret)", tag: "AUiRtmManager")
        if ret != 0 {
            completion(AUiCommonError.rtmError(ret).toNSError(), 0)
        }
    }
    
    func whoNow(channelName: String, completion:@escaping (Error?, [[String: String]]?)->()) {
        guard let presence = rtmClient.getPresence() else {
            completion(AUiCommonError.rtmError(-1).toNSError(), nil)
            return
        }
        
        let options = AgoraRtmPresenceOptions()
        options.includeUserId = true
        options.includeState = true
        let ret =
        presence.whoNow(channelName, channelType: kChannelType, options: options, completion: { resp, error in
            aui_info("presence whoNow '\(channelName)' finished: \(error.errorCode.rawValue) list count: \(resp.userStateList.count) userId: \(AUiRoomContext.shared.commonConfig?.userId ?? "")", tag: "AUiRtmManager")
            
            let userList = resp.userList()
            completion(error.toNSError(), userList)
        })
        aui_info("presence whoNow '\(channelName)' ret: \(ret)", tag: "AUiRtmManager")
        if ret != 0 {
            completion(AUiCommonError.rtmError(ret).toNSError(), nil)
        }
    }
    
    public func setPresenceState(channelName: String, attr:[String: Any], completion: @escaping (Error?)->()) {
        guard let presence = rtmClient.getPresence() else {
            completion(AUiCommonError.rtmError(-1).toNSError())
            return
        }
        
        var items: [AgoraRtmStateItem] = []
        attr.forEach { (key: String, value: Any) in
            let item = AgoraRtmStateItem()
            item.key = key
            if let val = value as? String {
                item.value = val
            } else if let val = value as? UInt {
                item.value = "\(val)"
            } else if let val = value as? Double {
                item.value = "\(val)"
            } else {
                aui_error("setPresenceState missmatch item: \(key): \(value)", tag: "AUiRtmManager")
                return
            }
            
            items.append(item)
        }
        let ret =
        presence.setState(channelName, channelType: kChannelType, items: items, completion: { resp, error in
            aui_info("presence setState '\(channelName)' finished: \(error.errorCode.rawValue)", tag: "AUiRtmManager")
            completion(error.toNSError())
        })
        aui_info("presence setState'\(channelName)' ret: \(ret)", tag: "AUiRtmManager")
        if ret != 0 {
            completion(AUiCommonError.rtmError(ret).toNSError())
        }
    }
}

//MARK: Channel Metadata
extension AUiRtmManager {
    public func subscribeMsg(channelName: String, itemKey: String, delegate: AUiRtmMsgProxyDelegate) {
        proxy.subscribeMsg(channelName: channelName, itemKey: itemKey, delegate: delegate)
    }
    
    public func unsubscribeMsg(channelName: String, itemKey: String, delegate: AUiRtmMsgProxyDelegate) {
        proxy.unsubscribeMsg(channelName: channelName, itemKey: itemKey, delegate: delegate)
    }
    
    public func subscribeUser(channelName: String, delegate: AUiRtmUserProxyDelegate) {
        proxy.subscribeUser(channelName: channelName, delegate: delegate)
    }
    
    public func unsubscribeUser(channelName: String, delegate: AUiRtmUserProxyDelegate) {
        proxy.unsubscribeUser(channelName: channelName, delegate: delegate)
    }
    
    public func subscribeError(channelName: String, delegate: AUiRtmErrorProxyDelegate) {
        proxy.subscribeError(channelName: channelName, delegate: delegate)
    }
    
    public func unsubscribeError(channelName: String, delegate: AUiRtmErrorProxyDelegate) {
        proxy.unsubscribeError(channelName: channelName, delegate: delegate)
    }
    
    func subscribe(channelName: String, rtcToken: String, completion:@escaping (Error?)->()) {
        if kChannelType == .message {
            let options = AgoraRtmSubscribeOptions()
            options.withMetadata = true
            options.withPresence = true
            let ret =
            rtmClient.subscribe(withChannel: channelName, option: options) { resp, error in
                aui_info("subscribe '\(channelName)' finished: \(error.errorCode.rawValue)", tag: "AUiRtmManager")
                completion(error.toNSError())
            }
            aui_info("subscribe '\(channelName)' ret: \(ret)", tag: "AUiRtmManager")
            if ret != 0 {
                completion(AUiCommonError.rtmError(ret).toNSError())
            }
        } else if kChannelType == .stream {
            let option = AgoraRtmJoinChannelOption()
            option.token = rtcToken
            option.withMetadata = true
            option.withPresence = true
            if rtmStreamChannelMap[channelName] == nil {
                let streamChannel = rtmClient.createStreamChannel(channelName)
                rtmStreamChannelMap[channelName] = streamChannel
                
            }
            guard let streamChannel = rtmStreamChannelMap[channelName] else {
                assert(false, "streamChannel not found")
                return
            }
            
            let ret = streamChannel.join(with: option) { resp, error in
                aui_info("join '\(channelName)' finished: \(error.errorCode.rawValue)", tag: "AUiRtmManager")
                completion(error.toNSError())
            }
            aui_info("join '\(channelName)' rtcToken: \(rtcToken) ret: \(ret)", tag: "AUiRtmManager")
            if ret != 0 {
                completion(AUiCommonError.rtmError(ret).toNSError())
            }
        } else {
            assert(false, "unknown channel type")
        }
    }
    
    func unSubscribe(channelName: String) {
        proxy.cleanCache(channelName: channelName)
        if kChannelType == .message {
            let ret =
            rtmClient.unsubscribe(withChannel: channelName)
            aui_info("unSubscribe '\(channelName)' ret: \(ret)", tag: "AUiRtmManager")
        } else if kChannelType == .stream {
            guard let streamChannel = rtmStreamChannelMap[channelName] else {
                return
            }
            
            streamChannel.leave()
            rtmStreamChannelMap[channelName] = nil
        }
    }
    
//    func cleanMetadata(channelName: String, completion: @escaping (Error?)->()) {
//        guard let data = rtmClient.getStorage()?.createMetadata(), let storage = rtmClient.getStorage() else {
//            assert(false, "cleanMetadata fail")
//            return
//        }
//        let options = AgoraRtmMetadataOptions()
//        options.recordTs = true
//        options.recordUserId = true
//        let ret =
//        storage.removeChannelMetadata(channelName, channelType: kChannelType, data: data, options: options, lock: "", completion: { resp, error in
//            aui_info("cleanMetadata finished: \(error.errorCode.rawValue)", tag: "AUiRtmManager")
//            completion(nil)
//        })
//        aui_info("cleanMetadata '\(channelName)' ret: \(ret)", tag: "AUiRtmManager")
//
//    }
//
//    func setMetadata(channelName: String, metadata: [String: String], completion: @escaping (Error?)->()) {
//        guard let storage = rtmClient.getStorage(),
//              let data = storage.createMetadata() else {
//            assert(false, "setMetadata fail")
//            return
//        }
//        metadata.forEach { (key: String, value: String) in
//            let item = AgoraRtmMetadataItem()
//            item.key = key
//            item.value = value
//            data.setMetadataItem(item)
//        }
//
//        let options = AgoraRtmMetadataOptions()
//        let ret =
//        storage.setChannelMetadata(channelName, channelType: kChannelType, data: data, options: options, lock: "", completion: { resp, error in
//            aui_info("setMetadata finished: \(error.errorCode.rawValue)", tag: "AUiRtmManager")
//            completion(nil)
//        })
//        aui_info("setMetadata ret: \(ret) \(metadata)", tag: "AUiRtmManager")
//    }
//
//    func updateMetadata(channelName: String, metadata: [String: String], completion: @escaping (Error?)->()) {
//        guard let storage = rtmClient.getStorage(),
//                  let data = storage.createMetadata() else {
//            assert(false, "updateMetadata fail")
//            return
//        }
//        metadata.forEach { (key: String, value: String) in
//            let item = AgoraRtmMetadataItem()
//            item.key = key
//            item.value = value
//            data.setMetadataItem(item)
//        }
//
//        let options = AgoraRtmMetadataOptions()
//        let ret =
//        storage.updateChannelMetadata(channelName, channelType: kChannelType, data: data, options: options, lock: "", completion: { resp, error in
//            aui_info("updateMetadata finished: \(error.errorCode.rawValue)", tag: "AUiRtmManager")
//            completion(nil)
//        })
//        aui_info("updateMetadata ret: \(ret)", tag: "AUiRtmManager")
//    }
    
    func getMetadata(channelName: String, completion: @escaping (Error?, [String: String]?)->()) {
        guard let storage = rtmClient.getStorage() else {
            assert(false, "getMetadata fail")
            return
        }
        let ret =
        storage.getChannelMetadata(channelName, channelType: kChannelType, completion: { resp, error in
            aui_info("getMetadata finished: \(error.errorCode.rawValue) item count: \(resp.data?.getItems().count ?? 0)", tag: "AUiRtmManager")
            var map: [String: String] = [:]
            resp.data?.getItems().forEach({ item in
                map[item.key] = item.value
            })
            completion(nil, map)
        })
        aui_info("getMetadata ret: \(ret)", tag: "AUiRtmManager")
    }
}

//MARK: user metadata
extension AUiRtmManager {
    func subscribeUser(userId: String) {
        guard let storage = rtmClient.getStorage() else {
            assert(false, "subscribeUserMetadata fail")
            return
        }
        let ret =
        storage.subscribeUserMetadata(userId, completion: { resp, error in
            aui_info("subscribeUser finished: \(resp) \(error.errorCode.rawValue)", tag: "AUiRtmManager")
        })
        aui_info("subscribeUserMetadata ret: \(ret)", tag: "AUiRtmManager")
    }
    
    func unSubscribeUser(userId: String) {
        guard let storage = rtmClient.getStorage() else {
            aui_error("subscribeUserMetadata fail", tag: "AUiRtmManager")
            assert(false, "subscribeUserMetadata fail")
            return
        }
        let ret =
        storage.unsubscribeUserMetadata(userId)
        aui_info("subscribeUserMetadata ret: \(ret)", tag: "AUiRtmManager")
    }
    
    func removeUserMetadata(userId: String) {
        guard let storage = rtmClient.getStorage(),
                  let data = storage.createMetadata() else {
            aui_info("removeUserMetadata fail", tag: "AUiRtmManager")
            assert(false, "removeUserMetadata fail")
            return
        }
        let options = AgoraRtmMetadataOptions()
        options.recordTs = true
        options.recordUserId = true
        
        let ret =
        storage.removeUserMetadata(userId, data: data, options: options, completion: { resp, error in
            aui_info("removeUserMetadata finished: \(error.errorCode.rawValue)", tag: "AUiRtmManager")
        })
        aui_info("removeUserMetadata ret: \(ret)", tag: "AUiRtmManager")
    }
    
    func setUserMetadata(userId: String, metadata: [String: String]) {
        guard let storage = rtmClient.getStorage(),
                let data = storage.createMetadata() else {
            assert(false, "setUserMetadata fail")
            return
        }
        let options = AgoraRtmMetadataOptions()
        options.recordTs = true
        options.recordUserId = true
        
        metadata.forEach { (key: String, value: String) in
            let item = AgoraRtmMetadataItem()
            item.key = key
            item.value = value
            data.setMetadataItem(item)
        }
        
        let ret =
        storage.setUserMetadata(userId, data: data, options: options, completion: { resp, error in
            aui_info("setUserMetadata finished: \(error.errorCode.rawValue)", tag: "AUiRtmManager")
        })
        aui_info("setUserMetadata ret: \(ret)", tag: "AUiRtmManager")
    }
    
    func updateUserMetadata(userId: String, metadata: [String: String]) {
        guard let storage = rtmClient.getStorage(),
                let data = storage.createMetadata() else {
            aui_error("updateUserlMetadata fail", tag: "AUiRtmManager")
            assert(false, "updateUserlMetadata fail")
            return
        }
        let options = AgoraRtmMetadataOptions()
        options.recordTs = true
        options.recordUserId = true
        
        metadata.forEach { (key: String, value: String) in
            let item = AgoraRtmMetadataItem()
            item.key = key
            item.value = value
            data.setMetadataItem(item)
        }
        
        let ret =
        storage.updateUserMetadata(userId, data: data, options: options, completion: { resp, error in
            aui_info("updateUserlMetadata finished: \(error.errorCode.rawValue)", tag: "AUiRtmManager")
        })
        aui_info("updateUserlMetadata ret: \(ret)", tag: "AUiRtmManager")
    }
    
    func getUserMetadata(userId: String) {
        guard let storage = rtmClient.getStorage() else {
            aui_error("getUserMetadata fail", tag: "AUiRtmManager")
            return
        }
        
        let ret =
        storage.getUserMetadata(userId) { resp, error in
            aui_info("getUserMetadata: \(resp) \(error.errorCode.rawValue)", tag: "AUiRtmManager")
        }
        aui_info("getUserMetadata ret: \(ret)", tag: "AUiRtmManager")
    }
}

File: Android/auikit/src/main/java/io/agora/auikit/service/rtm/AUiRtmManager.kt
package io.agora.auikit.service.rtm

import android.content.Context
import io.agora.auikit.utils.AUiLogger
import io.agora.rtm.ErrorInfo
import io.agora.rtm.JoinChannelOptions
import io.agora.rtm.Metadata
import io.agora.rtm.MetadataItem
import io.agora.rtm.MetadataOptions
import io.agora.rtm.PresenceOptions
import io.agora.rtm.ResultCallback
import io.agora.rtm.RtmClient
import io.agora.rtm.RtmConstants
import io.agora.rtm.RtmConstants.RtmChannelType
import io.agora.rtm.RtmEventListener
import io.agora.rtm.StateItem
import io.agora.rtm.StreamChannel
import io.agora.rtm.SubscribeOptions
import io.agora.rtm.WhoNowResult

class AUiRtmManager(
    context: Context,
    private val rtmClient: RtmClient,
) {

    private val proxy = AUiRtmMsgProxy()
    private val rtmStreamChannelMap = mutableMapOf<String, StreamChannel>()
    private val logger = AUiLogger(AUiLogger.Config(context, "AUiRtmManager"))
    @Volatile
    private var isLogin = false

    init {
        hookRtmEventListener()
    }

    private fun hookRtmEventListener() {
        val rtmClientClazz = rtmClient.javaClass
        val rtmEventListenerField = rtmClientClazz.getDeclaredField("mRtmEventListener")
        rtmEventListenerField.isAccessible = true
        val rtmEventListener = rtmEventListenerField.get(rtmClient) as RtmEventListener
        proxy.originEventListeners = rtmEventListener
        rtmEventListenerField.set(rtmClient, proxy)
    }

    // Channel Metadata
    private val kChannelType = RtmChannelType.STREAM


    fun login(token:String, completion: (AUiRtmException?) -> Unit){
        if(isLogin){
            completion.invoke(null)
            return
        }
        rtmClient.login(token, object : ResultCallback<Void>{
            override fun onSuccess(responseInfo: Void?) {
                isLogin = true
                completion.invoke(null)
            }

            override fun onFailure(errorInfo: ErrorInfo?) {
                if(errorInfo?.errorCode == RtmConstants.RTM_ERR_ALREADY_LOGIN){
                    isLogin = true
                    completion.invoke(null)
                }else{
                    completion.invoke(
                        AUiRtmException(
                        errorInfo?.errorCode ?: -1,
                        errorInfo?.errorReason ?: "UnKnow",
                        errorInfo?.operation ?: "UnKnow",
                    ))
                }
            }
        })
    }

    fun logout(){
        rtmClient.logout(object : ResultCallback<Void>{
            override fun onSuccess(responseInfo: Void?) {

            }

            override fun onFailure(errorInfo: ErrorInfo?) {

            }
        })
        isLogin = false
    }

    fun subscribeMsg(channelName: String, itemKey: String, delegate: AUiRtmMsgProxyDelegate) {
        proxy.subscribeMsg(channelName, itemKey, delegate)
    }

    fun unsubscribeMsg(channelName: String, itemKey: String, delegate: AUiRtmMsgProxyDelegate) {
        proxy.unsubscribeMsg(channelName, itemKey, delegate)
    }

    fun subscribeUser(delegate: AUiRtmUserProxyDelegate) {
        proxy.subscribeUser(delegate)
    }

    fun unsubscribeUser(delegate: AUiRtmUserProxyDelegate) {
        proxy.unsubscribeUser(delegate)
    }

    fun subscribe(channelName: String, token: String, completion: (AUiRtmException?) -> Unit) {
        when (kChannelType) {
            RtmChannelType.MESSAGE -> {
                val option = SubscribeOptions()
                option.withMetadata = true
                option.withPresence = true
                rtmClient.subscribe(channelName, option, object : ResultCallback<Void> {
                    override fun onSuccess(responseInfo: Void?) {
                        completion.invoke(null)
                    }

                    override fun onFailure(errorInfo: ErrorInfo?) {
                        if (errorInfo != null) {
                            completion.invoke(
                                AUiRtmException(errorInfo.errorCode, errorInfo.errorReason, errorInfo.operation)
                            )
                        } else {
                            completion.invoke(AUiRtmException(-1, "error", ""))
                        }
                    }
                })
            }
            RtmChannelType.STREAM -> {
                val option = JoinChannelOptions()
                option.token = token
                option.withMetadata = true
                option.withPresence = true
                if (rtmStreamChannelMap[channelName] == null) {
                    logger.d("EnterRoom", "create and join stream channel ...")
                    val streamChannel = rtmClient.createStreamChannel(channelName)
                    logger.d("StreamChannel", "joining... channelName=$channelName, token=$token")
                    streamChannel.join(option, object : ResultCallback<Void> {
                        override fun onSuccess(responseInfo: Void?) {
                            logger.d("EnterRoom", "create and join the stream channel successfully")
                            logger.d("StreamChannel", "join success channelName=$channelName")
                            completion.invoke(null)
                        }

                        override fun onFailure(errorInfo: ErrorInfo?) {
                            logger.d("EnterRoom", "create and join the stream channel failed for $errorInfo")
                            if (errorInfo != null) {
                                completion.invoke(AUiRtmException(errorInfo.errorCode, errorInfo.errorReason, errorInfo.operation))
                            } else {
                                completion.invoke(AUiRtmException(-1, "error", ""))
                            }
                        }
                    })
                    rtmStreamChannelMap[channelName] = streamChannel
                } else {
                    logger.d("EnterRoom", "create and join the stream channel failed for existing")
                    completion.invoke(
                        AUiRtmException(
                            -999,
                            "error for streamChannel existing",
                            "subscribe join stream channel"
                        )
                    )
                }
            }
            else -> {
                completion.invoke(AUiRtmException(-1, "error", ""))
            }
        }
    }

    fun unSubscribe(channelName: String) {
        proxy.cleanCache(channelName)
        when (kChannelType) {
            RtmChannelType.MESSAGE -> {
                rtmClient.unsubscribe(channelName)
            }
            RtmChannelType.STREAM -> {
                val streamChannel = rtmStreamChannelMap[channelName] ?: return
                streamChannel.leave(object : ResultCallback<Void> {
                    override fun onSuccess(responseInfo: Void?) {

                    }

                    override fun onFailure(errorInfo: ErrorInfo?) {

                    }
                })
                rtmStreamChannelMap.remove(channelName)
            }
        }
    }

    fun cleanMetadata(channelName: String, completion: (AUiRtmException?) -> Unit) {
        val storage = rtmClient.storage
        val data = storage.createMetadata()
        val options = MetadataOptions()
        options.recordTs = true
        options.recordUserId = true
        storage.removeChannelMetadata(channelName, kChannelType, data, options,
            "", object : ResultCallback<Void> {
                override fun onSuccess(responseInfo: Void?) {
                    completion.invoke(null)
                }

                override fun onFailure(errorInfo: ErrorInfo?) {
                    errorInfo ?: return
                    completion.invoke(
                        AUiRtmException(
                            errorInfo.errorCode,
                            errorInfo.errorReason,
                            errorInfo.operation
                        )
                    )
                }
            })
    }

    fun setMetadata(
        channelName: String,
        metadata: Map<String, String>,
        completion: (AUiRtmException?) -> Unit
    ) {
        val storage = rtmClient.storage
        val data = storage.createMetadata()
        metadata.forEach { entry ->
            val item = MetadataItem()
            item.key = entry.key
            item.value = entry.value
            data.setMetadataItem(item)
        }

        val options = MetadataOptions()
        storage.setChannelMetadata(
            channelName,
            kChannelType,
            data,
            options,
            "",
            object : ResultCallback<Void> {
                override fun onSuccess(responseInfo: Void?) {
                    completion.invoke(null)
                }

                override fun onFailure(errorInfo: ErrorInfo?) {
                    errorInfo ?: return
                    completion.invoke(
                        AUiRtmException(
                            errorInfo.errorCode,
                            errorInfo.errorReason,
                            errorInfo.operation
                        )
                    )
                }
            })
    }

    fun updateMetadata(
        channelName: String,
        metadata: Map<String, String>,
        completion: (AUiRtmException?) -> Unit
    ) {
        val storage = rtmClient.storage
        val data = storage.createMetadata()
        metadata.forEach { entry ->
            val item = MetadataItem()
            item.key = entry.key
            item.value = entry.value
            data.setMetadataItem(item)
        }
        val options = MetadataOptions()
        storage.updateChannelMetadata(
            channelName,
            kChannelType,
            data,
            options,
            "",
            object : ResultCallback<Void> {
                override fun onSuccess(responseInfo: Void?) {
                    completion.invoke(null)
                }

                override fun onFailure(errorInfo: ErrorInfo?) {
                    errorInfo ?: return
                    completion.invoke(
                        AUiRtmException(
                            errorInfo.errorCode,
                            errorInfo.errorReason,
                            errorInfo.operation
                        )
                    )
                }
            })
    }

    fun getMetadata(
        channelName: String,
        completion: (AUiRtmException?, Map<String, String>?) -> Unit
    ) {
        val storage = rtmClient.storage
        storage.getChannelMetadata(channelName, kChannelType, object : ResultCallback<Metadata> {
            override fun onSuccess(responseInfo: Metadata?) {
                responseInfo ?: return
                val map = mutableMapOf<String, String>()
                responseInfo.metadataItems.forEach { item ->
                    map[item.key] = item.value
                }
                completion.invoke(null, map)
            }

            override fun onFailure(errorInfo: ErrorInfo?) {
                errorInfo ?: return
                completion.invoke(
                    AUiRtmException(
                        errorInfo.errorCode,
                        errorInfo.errorReason,
                        errorInfo.operation
                    ), null
                )
            }
        })
    }


    // room list
    private val kRoomListChannelName = "uikitRoomList"
    private val kRoomListKey = "uikitRoomList"

    fun getRoomList(completion: (AUiRtmException?, String?) -> Unit) {
        getMetadata(kRoomListChannelName) { error, resp ->
            val roomListStr = resp?.get(kRoomListKey)
            completion.invoke(error, roomListStr)
        }
    }

    fun updateRoomList(listStr: String) {
        setMetadata(kRoomListChannelName, mapOf(Pair(kRoomListKey, listStr))) {}
    }

    fun whoNow(
        channelName: String,
        completion: (AUiRtmException?, List<Map<String, String>>?) -> Unit
    ) {
        val presence = rtmClient.presence
        val options = PresenceOptions()
        options.includeUserId = true
        options.includeState = true
        presence.whoNow(channelName, kChannelType, options, object: ResultCallback<WhoNowResult>{
            override fun onSuccess(responseInfo: WhoNowResult?) {
                responseInfo?: return
                var userList = arrayListOf<Map<String, String>>()
                responseInfo.userStateList.forEach { user ->
                    val userMap = mutableMapOf<String, String>()
                    userMap["userId"] = user.userId
                    user.states.forEach { item ->
                        userMap[item.key] = item.value
                    }
                    userList.add(userMap)
                }
                completion.invoke(null, userList)
            }

            override fun onFailure(errorInfo: ErrorInfo?) {
                errorInfo ?: return
                completion.invoke(AUiRtmException(errorInfo.errorCode, errorInfo.errorReason, errorInfo.operation), null)
            }
        })
    }

    fun setPresenceState(
        channelName: String,
        attr: Map<String, Any>,
        completion: (AUiRtmException?) -> Unit
    ) {
        val presence = rtmClient.presence
        val items = ArrayList<StateItem>()
        attr.forEach { entry ->
            val item = StateItem()
            item.key = entry.key
            item.value = entry.value.toString()
            items.add(item)
        }
        logger.d("PresenceState", "Setting channelName=$channelName, kChannelType=$kChannelType, items=$items")
        presence.setState(channelName, kChannelType, items, object : ResultCallback<Void>{
            override fun onSuccess(responseInfo: Void?) {
                logger.d("PresenceState", "Setting successfully")
                completion.invoke(null)
            }

            override fun onFailure(errorInfo: ErrorInfo?) {
                logger.d("PresenceState", "Setting failure : $errorInfo")
                completion.invoke(AUiRtmException(errorInfo?.errorCode ?: -1, errorInfo?.errorReason ?: "UnKnow Error", errorInfo?.operation ?: "UnKnow Error"))
            }
        })
    }

    // user metadata
    fun subscribeUser(userId: String){
        val storage = rtmClient.storage
        storage.subscribeUserMetadata(userId, object : ResultCallback<Void>{
            override fun onSuccess(responseInfo: Void?) {

            }

            override fun onFailure(errorInfo: ErrorInfo?) {

            }
        })
    }

    fun unSubscribeUser(userId: String){
        val storage = rtmClient.storage
        storage.unsubscribeUserMetadata(userId)
    }

    fun removeUserMetadata(userId: String){
        val storage = rtmClient.storage
        val data = storage.createMetadata()
        val options = MetadataOptions()
        options.recordTs = true
        options.recordUserId = true
        storage.removeUserMetadata(userId, data, options, object: ResultCallback<Void>{
            override fun onSuccess(responseInfo: Void?) {

            }

            override fun onFailure(errorInfo: ErrorInfo?) {

            }
        })
    }

    fun setUserMetadata(userId: String, metadata: Map<String, String>){
        val storage = rtmClient.storage
        val data = storage.createMetadata()
        val options = MetadataOptions()
        options.recordTs = true
        options.recordUserId = true
        metadata.forEach {entry ->
            val item = MetadataItem()
            item.key = entry.key
            item.value = entry.value
            data.setMetadataItem(item)
        }

        storage.setUserMetadata(userId, data, options, object: ResultCallback<Void>{
            override fun onSuccess(responseInfo: Void?) {

            }

            override fun onFailure(errorInfo: ErrorInfo?) {

            }
        })

    }

    fun updateUserMetadata(userId: String, metadata: Map<String, String>){
        val storage = rtmClient.storage
        val data = storage.createMetadata()
        val options = MetadataOptions()
        options.recordTs = true
        options.recordUserId = true
        metadata.forEach {entry ->
            val item = MetadataItem()
            item.key = entry.key
            item.value = entry.value
            data.setMetadataItem(item)
        }

        storage.updateUserMetadata(userId, data, options, object: ResultCallback<Void>{
            override fun onSuccess(responseInfo: Void?) {

            }

            override fun onFailure(errorInfo: ErrorInfo?) {

            }
        })
    }

    fun getUserMetadata(userId: String){
        val storage = rtmClient.storage
        storage.getUserMetadata(userId, object : ResultCallback<Metadata>{
            override fun onSuccess(responseInfo: Metadata?) {

            }

            override fun onFailure(errorInfo: ErrorInfo?) {

            }
        })
    }

}